src/arch/x86/x86_64/src/bootblock.S

```diff
@@ -0,0 +1,242 @@
+	 *
+	 * The linker scripts computes gdtptr16_offset by simply returning
+	 * the low 16 bits.  This means that the initial segment used
+	 * when start is called must be 64K aligned.  This should not
+	 * restrict the address as the ip address can be anything.
+	 *
+	 * Also load an IDT with NULL limit to prevent the 16bit IDT being used
+	 * in protected mode before c_start.S sets up a 32bit IDT when entering
+	 * RAM stage. In practise: CPU will shutdown on any exception.
+	 * See IA32 manual Vol 3A 19.26 Interrupts.
+	 */
+
+/*
+	movw	%cs, %ax
+	shlw	$4, %ax
+	movw	$nullidt_offset, %bx
+	subw	%ax, %bx
+	lidt	%cs:(%bx)
+	movw	$gdtptr16_offset, %bx
+	subw	%ax, %bx
+	lgdtl	%cs:(%bx)
+*/
+
+	movl	%cr0, %eax
+	andl	$0x7FFAFFD1, %eax /* PG,AM,WP,NE,TS,EM,MP = 0 */
+	orl	$0x60000001, %eax /* CD, NW, PE = 1 */
+	movl	%eax, %cr0
+
+	/* Restore BIST to %eax */
+	movl	%ebp, %eax
+
+	/* Now that we are in protected mode jump to a 32 bit code segment. */
+	ljmpl	$ROM_CODE_SEG, $__protected_start
+
+	// TODO: should set accessed and dirty bits in gdt entries
+	// so CPU does not try to write them to ROM?
+.align	4
+.code32
+.section ".text._gdt_", "ax", @progbits
+
+	.globl gdt_init
+gdt_init:
+	lgdt	%cs:gdtptr
+	ret
+
+.previous
+	.align	4
+.globl gdtptr
+gdtptr:
+	.word	gdt_end - gdt -1 /* compute the table limit */
+	.long	gdt		 /* we know the offset */
+
+	.align	4
+gdt:
+	/* selgdt 0, unused */
+	.word	0x0000, 0x0000		/* dummy */
+	.byte	0x00, 0x00, 0x00, 0x00
+
+	/* selgdt 0x08, flat code segment */
+	.word	0xffff, 0x0000
+	.byte	0x00, 0x9b, 0xcf, 0x00 /* G=1 and 0x0f, So we get 4Gbytes
+					  for limit */
+
+	/* selgdt 0x10,flat data segment */
+	.word	0xffff, 0x0000
+	.byte	0x00, 0x93, 0xcf, 0x00
+
+	/* selgdt 0x18, flat code segment (64-bit) */
+	.word   0xffff, 0x0000
+	.byte   0x00, 0x9b, 0xaf, 0x00
+
+gdt_end:
+.globl gdtptr16
+gdtptr16:
+	.word	gdt_end - gdt -1 /* compute the table limit */
+	.long	gdt		 /* we know the offset */
+
+.align	4
+.globl nullidt
+nullidt:
+	.word	0	/* limit */
+	.long	0
+	.word	0
+
+.globl _estart16bit
+_estart16bit:
+	.code32
+
+	// we're now in 32-bit mode.
+/*
+ *	When we come here we are in protected mode. We expand
+ *	the stack and copies the data segment from ROM to the
+ *	memory.
+ *
+ *	After that, we call the chipset bootstrap routine that
+ *	does what is left of the chipset initialization.
+ *
+ *	NOTE aligned to 4 so that we are sure that the prefetch
+ *	cache will be reloaded.
+ *
+ *	In the bootblock there is already a ljmp to __protected_start and
+ *	the reset vector jumps to symbol _start16bit in entry16.inc from
+ *	the reset vectors's symbol which is _start. Therefore, don't
+ *	expose the _start symbol for bootblock.
+ */
+	.align	4
+.globl _start32
+_start32:
+
+	lgdt	%cs:gdtptr
+	ljmp	$ROM_CODE_SEG, $__protected_start
+
+__protected_start:
+	/* Save the BIST value */
+	movl	%eax, %ebp
+
+	//post_code(POST_ENTER_PROTECTED_MODE)
+
+
+	movw	$ROM_DATA_SEG, %ax
```

Comment:
With this, I get the link error "oreboot.184q9sr1-cgu.0:(.text+0x73): undefined reference to `ROM_DATA_SEG'".